File: Source/ApplicationModel/Frontend/commands/CommandResult.ts
// Copyright (c) Aksio Insurtech. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import { Guid } from '@aksio/cratis-fundamentals';
import { ICommandResult } from './ICommandResult';
import { ValidationResult } from '../validation/ValidationResult';
import { Constructor, JsonSerializer } from '@aksio/cratis-fundamentals';

/**
 * Represents the result from executing a {@link ICommand}.
 */
export class CommandResult<TResponse = {}> implements ICommandResult<TResponse> {

    static empty: CommandResult = new CommandResult({
        correlationId: Guid.empty.toString(),
        isSuccess: true,
        isAuthorized: true,
        isValid: true,
        hasExceptions: false,
        validationErrors: [],
        exceptionMessages: [],
        exceptionStackTrace: '',
        response: null
    }, Object, false);

    /** @inheritdoc */
    readonly correlationId: Guid;

    /** @inheritdoc */
    readonly isSuccess: boolean;

    /** @inheritdoc */
    readonly isAuthorized: boolean;

    /** @inheritdoc */
    readonly isValid: boolean;

    /** @inheritdoc */
    readonly hasExceptions: boolean;

    /** @inheritdoc */
    readonly validationResults: ValidationResult[];

    /** @inheritdoc */
    readonly exceptionMessages: string[];

    /** @inheritdoc */
    readonly exceptionStackTrace: string;

    /** @inheritdoc */
    readonly response?: TResponse;

    /**
     * Creates an instance of command result.
     * @param {*} result The JSON/any representation of the command result;
     * @param {Constructor} responseInstanceType The {@see Constructor} that represents the type of response, if any. Defaults to {@see Object}.
     * @param {boolean} isResponseTypeEnumerable Whether or not the response type is an enumerable or not.
     */
    constructor(result: any, responseInstanceType: Constructor = Object, isResponseTypeEnumerable: boolean) {
        this.correlationId = Guid.parse(result.correlationId);
        this.isSuccess = result.isSuccess;
        this.isAuthorized = result.isAuthorized;
        this.isValid = result.isValid;
        this.hasExceptions = result.hasExceptions;
        this.validationResults = result.validationErrors.map(_ => new ValidationResult(_.severity, _.message, _.members, _.state));
        this.exceptionMessages = result.exceptionMessages;
        this.exceptionStackTrace = result.exceptionStackTrace;

        if (result.response) {
            if (isResponseTypeEnumerable) {
                this.response = JsonSerializer.deserializeArrayFromInstance(responseInstanceType, result.response) as any;
            } else {
                this.response = JsonSerializer.deserializeFromInstance(responseInstanceType, result.response) as any;
            }
        }
    }
}

File: Source/Fundamentals/Commands/CommandResult.cs
// Copyright (c) Aksio Insurtech. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using Aksio.Cratis.Validation;
using Aksio.Cratis.Execution;

namespace Aksio.Cratis.Commands;

/// <summary>
/// Represents the result coming from executing a command.
/// </summary>
public class CommandResult
{
    /// <summary>
    /// Represents a successful command result.
    /// </summary>
    public static readonly CommandResult Success = new();

    /// <summary>
    /// Gets the <see cref="CorrelationId"/> associated with the command.
    /// </summary>
    public CorrelationId CorrelationId { get; init; } = new(Guid.Empty.ToString());

    /// <summary>
    /// Gets whether or not the command executed successfully.
    /// </summary>
    public bool IsSuccess => IsAuthorized && IsValid && !HasExceptions;

    /// <summary>
    /// Gets whether or not the command was authorized to execute.
    /// </summary>
    public bool IsAuthorized { get; init; } = true;

    /// <summary>
    /// Gets whether or not the command is valid.
    /// </summary>
    public bool IsValid => !ValidationErrors.Any();

    /// <summary>
    /// Gets whether or not there are any exceptions that occurred.
    /// </summary>
    public bool HasExceptions => ExceptionMessages.Any();

    /// <summary>
    /// Gets any validation errors. If this collection is empty, there are errors.
    /// </summary>
    public IEnumerable<ValidationError> ValidationErrors { get; init; } = Enumerable.Empty<ValidationError>();

    /// <summary>
    /// Gets any exception messages that might have occurred.
    /// </summary>
    public IEnumerable<string> ExceptionMessages { get; init; } = Enumerable.Empty<string>();

    /// <summary>
    /// Gets the stack trace if there was an exception.
    /// </summary>
    public string ExceptionStackTrace { get; init; } = string.Empty;

    /// <summary>
    /// Optional response object. Controller actions representing a command can optionally return a response as any type, this is where it would be.
    /// </summary>
    public object? Response { get; init; }
}
